File: p2/p211/WordDictionary.java
package leetcode.p2.p211;

import java.util.LinkedList;
import java.util.Queue;

public class WordDictionary {

    private TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }

    public void addWord(String word) {
        TrieNode cur = root;
        for (int i = 0; i < word.length(); i++) {
            int index = word.charAt(i) - 'a';
            if (cur.children[index] == null) {
                cur.children[index] = new TrieNode();
            }
            cur = cur.children[index];
        }
        cur.isWord = true;
    }

    public boolean search(String word) {
        Queue<TrieNode> queue = new LinkedList<>();
        queue.offer(root);
        int index = 0;
        while (!queue.isEmpty() && index <= word.length()) {
            Queue<TrieNode> nextQueue = new LinkedList<>();
            while (!queue.isEmpty()) {
                TrieNode cur = queue.poll();
                if (index < word.length()) {
                    if (word.charAt(index) == '.') {
                        for (int i = 0; i < cur.children.length; i++) {
                            if (cur.children[i] != null) {
                                nextQueue.offer(cur.children[i]);
                            }
                        }
                    } else {
                        int temp = word.charAt(index) - 'a';
                        if (cur.children[temp] != null) {
                            nextQueue.offer(cur.children[temp]);
                        }
                    }
                } else {
                    if (cur.isWord) {
                        return true;
                    }
                }
            }
            queue = nextQueue;
            index++;
        }
        return false;
    }
}


class TrieNode {
    boolean isWord;
    TrieNode[] children;

    public TrieNode() {
        isWord = false;
        children = new TrieNode[26];
    }
}

File: p2/p211/word_dictionary.py
"""
Design a data structure that supports adding new words and finding if a string matches any previously added string.

Implement the WordDictionary class:

WordDictionary() Initializes the object.
void addWord(word) Adds word to the data structure, it can be matched later.
bool search(word) Returns true if there is any string in the data structure that matches word or false otherwise. word may contain dots '.' where dots can be matched with any letter.


Example:

Input
["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
[[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
Output
[null,null,null,null,false,true,true,true]

Explanation
WordDictionary wordDictionary = new WordDictionary();
wordDictionary.addWord("bad");
wordDictionary.addWord("dad");
wordDictionary.addWord("mad");
wordDictionary.search("pad"); // return False
wordDictionary.search("bad"); // return True
wordDictionary.search(".ad"); // return True
wordDictionary.search("b.."); // return True


Constraints:

1 <= word.length <= 25
word in addWord consists of lowercase English letters.
word in search consist of '.' or lowercase English letters.
There will be at most 2 dots in word for search queries.
At most 104 calls will be made to addWord and search.
"""


class WordDictionary:
    """
    Jun 30, 2025 16:09
    """
    class TrieNode:
        def __init__(self):
            self.is_word = False
            self.children = {}

    def __init__(self):
        self.root = self.TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = self.TrieNode()
            node = node.children[c]

        node.is_word = True

    def search(self, word: str) -> bool:
        return self.search_helper(self.root, word, 0)

    def search_helper(self, node: TrieNode, word, index) -> bool:
        if index == len(word):
            return node.is_word

        if word[index] == '.':
            for c in node.children:
                if self.search_helper(node.children[c], word, index + 1):
                    return True
            return False
        else:
            if word[index] not in node.children:
                return False

            return self.search_helper(node.children[word[index]], word, index + 1)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

if __name__ == '__main__':
    obj = WordDictionary()
    obj.addWord("a")
    obj.addWord("a")
    print(obj.search("."))
    print(obj.search("a"))
    print(not obj.search("aa"))
    print(obj.search("a"))
    print(obj.search(".a") == False)
    print(obj.search("a.") == False)

    obj = WordDictionary()
    obj.addWord("bad")
    obj.addWord("dad")
    obj.addWord("mad")
    print(not obj.search("pad"))
    print(obj.search("bad"))
    print(obj.search(".ad"))
    print(obj.search("b.."))
